Sources/BraintreeCore/RepeatingTimer.swift

```diff
@@ -0,0 +1,71 @@
+import Foundation
+
+final class RepeatingTimer {
+    
+    /// Amount of time, in seconds.
+    private let timeInterval: Int
+    
+    init(timeInterval: Int) {
+        self.timeInterval = timeInterval
+    }
+    
+    private lazy var timer: DispatchSourceTimer = {
+        let timerSource = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
+        timerSource.schedule(
+            deadline: .now() + .seconds(timeInterval),
+            repeating: .seconds(timeInterval),
+            leeway: .seconds(1)
+        )
+        timerSource.setEventHandler { [weak self] in
+            self?.eventHandler?()
+        }
+        return timerSource
+    }()
+    
+    var eventHandler: (() -> Void)?
+    
+    private var state: State = .suspended
+    
+    private enum State {
+        case suspended
+        case resumed
+    }
+    
+    deinit {
+        timer.setEventHandler { }
+        timer.cancel()
+        // If the timer is suspended, calling cancel without resuming afterwards
+        // triggers a crash. This is documented here https://forums.developer.apple.com/thread/15902
+        resume()
+        eventHandler = nil
+    }
+    
+    // MARK: - Public Methods
+    
+    /*
+     GCD timers are sensitive to errors. It is crucial to maintain
+     balance between calls to `dispatch_suspend` and `dispatch_resume`. Failure to do so
+     results in crashes with errors similar to:
+
+       BUG IN CLIENT OF LIBDISPATCH: Over-resume of an object
+
+     Such errors indicate an attempt to resume an already resumed timer. According to the
+     documentation, each call to `dispatch_suspend` must be matched with a corresponding
+     call to `dispatch_resume` to ensure proper event delivery and avoid crashes.
+     For more information: https://developer.apple.com/library/archive/documentation/General/Conceptual/ConcurrencyProgrammingGuide/GCDWorkQueues/GCDWorkQueues.html#//apple_ref/doc/uid/TP40008091-CH103-SW8
+    */
+    
+    func resume() {
```

Comment:
This PR is great work Rich. I am wondering if we want to synchronize resume and suspends as well since analytics could be sent from different threads. We could use actor but it would require waits on function calls and we could also use serial dispatch queue. wdyt?

Reply:
Excellent suggestion. For now, I think `DispatchSource` is a task that runs periodically, so we could proceed without needing _actor_'s. Also, based on Jax's implementation, the `DispatchSource` is created to run on the main thread. From what I see, the task of sending analytics can be on different threads, but both only enqueue events. The task of sending them is handled by the single instance of `DispatchSource`.

Reply:
Interesting. I didn't know that this line `let timerSource = DispatchSource.makeTimerSource(queue: DispatchQueue.main)` ensures that the handler is run on main as well.